numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +336,93 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!intern_strings())
+        return NULL;
```

Comment:
We could possibly also just add them to the `intern_strings` function in `multiarray.c` its probably no worse than here, and at least they are all in one place...

Reply:
No, you're not wrong. Given `PyUnicode_InternFromString` will only create the object once I thought that would be an acceptable cost. Unlike other cases XXX and XXX, we don't have an initializer function for this file where it's guaranteed it will be only executed once. I think we could solve this with a static variable at the top of `arrayfunction_override.c` that would be set upon the first call of `intern_strings` to prevent calling it multiple times. What do you think of this idea? Maybe there are better solutions, let me know if you have something in mind.

Thanks, I wasn't sure if it was part of the formatting for one-liner `if`s, but I'll add them now.

Reply:
Yes, the cost probably is not very relevant. But I think moving it to `multiarray.c` and exporting (privately in NumPy) similar to the other strings there is just as well.  For example `__array_ufunc__` is interned there (`__array_function__` is not, because we use it in a way where interning is not as useful).

I expect, that currently the code also leaks references to the strings, since the string creation will return a new reference every time, so in that sense it might simplify things a bit.

Reply:
This is a great idea, I did so now in the latest commits, thanks for the suggestion!